### PR TITLE
Added RuntimeWarning to lib.fast_unique_multiple

### PR DESCRIPTION
**Pull request by ShaharNaveh**
- [ ] closes #xxxx
- [ ] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/_libs/lib.pyx [hunk 1]
```diff
@@ -286,7 +287,9 @@ def fast_unique_multiple(list arrays, sort: bool = True):
         try:
             uniques.sort()
         except TypeError:
-            # TODO: RuntimeWarning?
+            warnings.warn(
+                "Not sure what will be a helpful error message", RuntimeWarning
```
Comment (MomIsBestFriend) [inline pandas/_libs/lib.pyx]: I would love if someone could give me an idea, on what will be a helpful warning message here.

File: pandas/tests/test_lib.py [hunk 2]
```diff
@@ -39,6 +40,11 @@ def test_fast_unique_multiple_list_gen_sort(self):
         out = lib.fast_unique_multiple_list_gen(gen, sort=False)
         tm.assert_numpy_array_equal(np.array(out), expected)
 
+    def test_fast_unique_multiple_unsortable_runtimewarning(self):
+        idx = pd.MultiIndex.from_product([[1, pd.Timestamp("2000")], ["foo", "bar"]])
```
Comment (jbrockmendel) [inline pandas/tests/test_lib.py]: can you target this test more specifically at the affected function?  maybe pass `np.array([Timestamp(...), "foo"])`?

It's not a rule/policy, but ideally the intra-pandas dependencies of test_lib would be limited to the dependencies of `lib`.

File: pandas/_libs/lib.pyx (hunk 1 above)
Comment (jbrockmendel) [reply pandas/_libs/lib.pyx]: i know i suggested searching for TODO comments; i should have specified that the idea would to aim for ones that you _can_ solve, ideally that push your limits.  From a reviewer's perspective, this is pretty similar to the existing TODO comment

Comment (MomIsBestFriend) [reply pandas/_libs/lib.pyx]: > From a reviewer's perspective, this is pretty similar to the existing TODO comment

Haven't thought of that, I can see it now.

**Comment by MomIsBestFriend** [comment]
Looks like at ```L340```, it triggers the warning, but only at the AP ```Linux py37_np_dev```

https://github.com/pandas-dev/pandas/blob/42ef40971261b286b53ddfc85d2e399067b7c72d/pandas/tests/indexes/multi/test_setops.py#L334-L345

---

@jbrockmendel Does having something like this:

```python
if _is_numpy_dev:
    with tm.assert_produces_warning(RuntimeWarning):
        result = idx.union(idx[:1]) 
else:
    result = idx.union(idx[:1]) 
```

is the correct approach?

**Comment by jbrockmendel** [comment]
Usually we would write

```
warn = RuntimeWarning if _is_numpy_dev else None
with tm.assert_produces_warning(warn):
    ...
```

File: pandas/tests/test_lib.py [hunk 3]
```diff
@@ -39,6 +40,11 @@ def test_fast_unique_multiple_list_gen_sort(self):
         out = lib.fast_unique_multiple_list_gen(gen, sort=False)
         tm.assert_numpy_array_equal(np.array(out), expected)
 
+    def test_fast_unique_multiple_unsortable_runtimewarning(self):
+        arr = [np.array(["foo", pd.Timestamp("2000")])]
+        with tm.assert_produces_warning(RuntimeWarning):
+            lib.fast_unique_multiple(arr, sort=None)
```
Comment (WillAyd) [inline pandas/tests/test_lib.py]: Can you test this from an end user perspective?

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/tests/indexes/multi/test_setops.py [hunk 4]
```diff
@@ -337,7 +339,9 @@ def test_union_sort_other_incomparable():
     idx = pd.MultiIndex.from_product([[1, pd.Timestamp("2000")], ["a", "b"]])
 
     # default, sort=None
-    result = idx.union(idx[:1])
+    warn = None if _is_numpy_dev else RuntimeWarning
+    with tm.assert_produces_warning(warn):
+        result = idx.union(idx[:1])
```
Comment (MomIsBestFriend) [inline pandas/tests/indexes/multi/test_setops.py]: > Can you test this from an end user perspective?

@WillAyd Isn't this already testing from an end user perspective?

Comment (WillAyd) [reply pandas/tests/indexes/multi/test_setops.py]: Ah sorry overlooked this one

File: pandas/_libs/lib.pyx [hunk 5]
```diff
@@ -286,7 +287,7 @@ def fast_unique_multiple(list arrays, sort: bool = True):
         try:
             uniques.sort()
         except TypeError:
-            # TODO: RuntimeWarning?
+            warnings.warn("The values in the array are unorderable", RuntimeWarning)
```
Comment (jorisvandenbossche) [inline pandas/_libs/lib.pyx]: Can you indicate somethig about what to do about this? (eg passing the `sort=False` keyword? I don't know if that is possible for all cases that may end up here)

File: pandas/_libs/lib.pyx [hunk 6]
```diff
@@ -287,7 +287,11 @@ def fast_unique_multiple(list arrays, sort: bool = True):
         try:
             uniques.sort()
         except TypeError:
-            warnings.warn("The values in the array are unorderable", RuntimeWarning)
+            warnings.warn(
+                "The values in the array are unorderable. "
+                "Pass `sort=Flase` to suppress this warning.",
```
Comment (jorisvandenbossche) [inline pandas/_libs/lib.pyx]:
```suggestion
                "Pass `sort=False` to suppress this warning.",
```

File: pandas/tests/indexes/multi/test_setops.py [hunk 7]
```diff
@@ -332,12 +334,15 @@ def test_union_sort_other_empty_sort(slice_):
     tm.assert_index_equal(result, expected)
 
 
+@pytest.mark.xfail(reason="`warn` if statement is not working properly", strict=False)
```
Comment (jbrockmendel) [inline pandas/tests/indexes/multi/test_setops.py]: why is this necesssary?  non-strict xfails are among the patterns we want to avoid

Comment (MomIsBestFriend) [reply pandas/tests/indexes/multi/test_setops.py]: This is necessary because I couldn't get the CI to pass otherwise

Comment (jbrockmendel) [reply pandas/tests/indexes/multi/test_setops.py]: OK, lets troubleshoot that rather than xfail.

Why is there no expected warning on numpy-dev?

Comment (MomIsBestFriend) [reply pandas/tests/indexes/multi/test_setops.py]: I don't know, I have tried in d2cd2d9 to have:

```python
warn = None if _is_numpy_dev else RuntimeWarning
```

---

We can trouble shoot from there.

**Comment by MomIsBestFriend** [comment]
So apparently I never needed the ```_is_numpy_dev```. \*facepalm\*

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm. can you add a release note, can put in Other Enhancements for 1.1

File: doc/source/whatsnew/v1.1.0.rst [hunk 8]
```diff
@@ -87,6 +87,7 @@ Other enhancements
 - Positional slicing on a :class:`IntervalIndex` now supports slices with ``step > 1`` (:issue:`31658`)
 - :class:`Series.str` now has a `fullmatch` method that matches a regular expression against the entire string in each row of the series, similar to `re.fullmatch` (:issue:`32806`).
 - :meth:`DataFrame.sample` will now also allow array-like and BitGenerator objects to be passed to ``random_state`` as seeds (:issue:`32503`)
+- :meth:`MultiIndex.union` will now raise `RuntimeWarning` if the object inside are unsortable, pass `sort=False` to suppress this warning.
```
Comment (jreback) [inline doc/source/whatsnew/v1.1.0.rst]: can you add the issue number at the end (this PR number is fine if we don't have an issue)

**Review by jreback — CHANGES_REQUESTED** [review]
ping on green.

**Comment by MomIsBestFriend** [comment]
ping @jreback

**Review by jreback — APPROVED** [review]

File: pandas/tests/test_lib.py (hunk 3 above)
Comment (jreback) [reply pandas/tests/test_lib.py:46]: see above

**Comment by jreback** [comment]
thanks @MomIsBestFriend

**Pull request merged**